vault/plugin_reload.go

```diff
@@ -73,49 +73,101 @@ func (c *Core) reloadMatchingPluginMounts(ctx context.Context, mounts []string)
 // reloadPlugin reloads all mounted backends that are of
 // plugin pluginName (name of the plugin as registered in
 // the plugin catalog).
-func (c *Core) reloadMatchingPlugin(ctx context.Context, pluginName string) error {
-	c.mountsLock.RLock()
-	defer c.mountsLock.RUnlock()
-	c.authLock.RLock()
-	defer c.authLock.RUnlock()
+func (c *Core) reloadMatchingPlugin(ctx context.Context, pluginName string) (int, error) {
+	var reloaded int
+	typeExists := map[consts.PluginType]bool{
+		consts.PluginTypeCredential: false,
+		consts.PluginTypeDatabase:   false,
+		consts.PluginTypeSecrets:    false,
+	}
+	for pluginType := range typeExists {
+		plugins, err := c.pluginCatalog.ListVersionedPlugins(ctx, pluginType)
+		if err != nil {
+			return reloaded, err
+		}
+		for _, plugin := range plugins {
+			if plugin.Name == pluginName {
+				typeExists[pluginType] = true
+				break
+			}
```

Comment:
It's a bit odd, but this preserves the existing behaviour of the reload endpoint, which only refers to plugin by name and doesn't filter on type at all: https://developer.hashicorp.com/vault/api-docs/system/plugins-reload-backend#reload-plugins

That's actually part of why in my next PR I'm going to add a new endpoint instead of updating the existing one: https://github.com/hashicorp/vault/compare/vault-21610/root-plugin-reload-api?expand=1. The other reason for the new endpoint is that I want the new endpoint to only be available in the root namespace (because it is cross-namespace and designed to be paired with the root-namespace-only plugin catalog), but the existing reload endpoint gets mounted in all sub-namespaces too.